# ① Introduction
  과대적합은 딥러닝 모델이 훈련 데이터에만 특화된 패턴을 학습하기 시작했다는 의미입니다.  
  훈련 데이터에만 특화된 패턴은 새로운 데이터와 관련성이 적어 잘못된 판단을 하게 만듭니다.  
  과대적합을 해결하려면 더 많은 훈련 데이터를 모아야 합니다.  
  하지만 현실적으로 더 많은 훈련 데이터를 모으는 것이 불가능할 때 차선책은 모델이 저장할 수 있는 정보에 제약을 가하는 것입니다.   
  즉, 모델의 학습 파라미터 수를 줄이는 것입니다.   
  유튜브: https://youtu.be/zPJNdAG6T3s

# ② Methods

### Import

In [67]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adagrad

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import metrics

import pandas as pd
import numpy as np

### 샘플 개수
훈련 데이터의 샘플 개수는 141개로 작기 때문에 과대적합될 가능성이 높습니다.  
따라서 모델의 학습 파라미터 수를 줄여줍니다.

In [76]:
df = pd.read_csv('data.csv')
len(df)

141

In [69]:
df['수익률'] = df['수익률'].apply(lambda x : 1 if x > 0 else 0 )
df['수익률'].unique()

array([1, 0], dtype=int64)

In [70]:
X = df[['뉴스기사', '직전 거래대금', '최대거래대금', '양봉개수']]
y = df['수익률']

X = X.to_numpy()
y = y.to_numpy()

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

model = Sequential()
model.add(Dense(4, input_dim=4, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
optimizer = Adagrad(lr=0.2)

model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

C:\Users\user\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\user\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
C:\Users\user\Anaconda3\lib\site-packages\sklearn\utils\validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [71]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_34 (Dense)             (None, 4)                 20        
_________________________________________________________________
dense_35 (Dense)             (None, 4)                 20        
_________________________________________________________________
dense_36 (Dense)             (None, 1)                 5         
Total params: 45
Trainable params: 45
Non-trainable params: 0
_________________________________________________________________


In [72]:
model.fit(X_train, y_train, epochs=200, batch_size=40)

Epoch 1/200
105/105 [==============================] - 1s 6ms/step - loss: 0.6656 - acc: 0.6190
Epoch 2/200
105/105 [==============================] - 0s 48us/step - loss: 0.6734 - acc: 0.6000
Epoch 3/200
105/105 [==============================] - 0s 76us/step - loss: 0.6540 - acc: 0.6476
Epoch 4/200
105/105 [==============================] - 0s 57us/step - loss: 0.6255 - acc: 0.6571
Epoch 5/200
105/105 [==============================] - 0s 38us/step - loss: 0.6214 - acc: 0.6381
Epoch 6/200
105/105 [==============================] - 0s 48us/step - loss: 0.5931 - acc: 0.7048
Epoch 7/200
105/105 [==============================] - 0s 86us/step - loss: 0.5854 - acc: 0.7048
Epoch 8/200
105/105 [==============================] - 0s 86us/step - loss: 0.5744 - acc: 0.6762
Epoch 9/200
105/105 [==============================] - 0s 57us/step - loss: 0.5717 - acc: 0.6952
Epoch 10/200
105/105 [==============================] - 0s 105us/step - loss: 0.5639 - acc: 0.6952
Epoch 11/200
105/105 [=======

105/105 [==============================] - 0s 76us/step - loss: 0.5108 - acc: 0.7143
Epoch 85/200
105/105 [==============================] - 0s 67us/step - loss: 0.5172 - acc: 0.7238
Epoch 86/200
105/105 [==============================] - 0s 67us/step - loss: 0.5117 - acc: 0.7238
Epoch 87/200
105/105 [==============================] - 0s 57us/step - loss: 0.5075 - acc: 0.7238
Epoch 88/200
105/105 [==============================] - 0s 76us/step - loss: 0.5147 - acc: 0.7143
Epoch 89/200
105/105 [==============================] - 0s 48us/step - loss: 0.5096 - acc: 0.7238
Epoch 90/200
105/105 [==============================] - 0s 57us/step - loss: 0.5066 - acc: 0.7143
Epoch 91/200
105/105 [==============================] - 0s 38us/step - loss: 0.5163 - acc: 0.7238
Epoch 92/200
105/105 [==============================] - 0s 57us/step - loss: 0.5060 - acc: 0.7238
Epoch 93/200
105/105 [==============================] - 0s 67us/step - loss: 0.5105 - acc: 0.7238
Epoch 94/200
105/105 [===========

105/105 [==============================] - 0s 86us/step - loss: 0.4918 - acc: 0.7143
Epoch 167/200
105/105 [==============================] - 0s 67us/step - loss: 0.4921 - acc: 0.7238
Epoch 168/200
105/105 [==============================] - 0s 57us/step - loss: 0.4942 - acc: 0.7143
Epoch 169/200
105/105 [==============================] - 0s 67us/step - loss: 0.4926 - acc: 0.7143
Epoch 170/200
105/105 [==============================] - 0s 76us/step - loss: 0.4920 - acc: 0.7143
Epoch 171/200
105/105 [==============================] - 0s 57us/step - loss: 0.4924 - acc: 0.7238
Epoch 172/200
105/105 [==============================] - 0s 67us/step - loss: 0.4929 - acc: 0.7143
Epoch 173/200
105/105 [==============================] - 0s 67us/step - loss: 0.4940 - acc: 0.7143
Epoch 174/200
105/105 [==============================] - 0s 67us/step - loss: 0.4920 - acc: 0.7143
Epoch 175/200
105/105 [==============================] - 0s 48us/step - loss: 0.4935 - acc: 0.7143
Epoch 176/200
105/105 [=

In [73]:
_, accuracy = model.evaluate(X_test, y_test)
print(accuracy)

36/36 [==============================] - 0s 6ms/step
0.6111111111111112


# ③ Results
Layer가 12, 8, 1 일 때 Score는 55%였는데  
Layer가 4, 4, 1일 때 Score는 61%로 개선되었다.

# ④ Conclusion
학습 파라미터를 줄이니 과대적합이 다소 해소되었다.